### PR TITLE
Fix ECAL Endcap quadrants issue with DD4hep reflected volumes + Also fix ECAL Barrel and CSCs reflected active detectors

### PR DESCRIPTION
**Pull request by ghugo83**
**This PR fixes the handling of reflected volumes in CMSSW in DD4hep workflows.**

The key idea is that the handling of reflected volumes is different in old DD and DD4hep, as there are 2 different files in CMSSW:
- old DD: https://github.com/cms-sw/cmssw/blob/master/SimG4Core/Geometry/src/DDG4Builder.cc 
- DD4hep: https://github.com/cms-sw/cmssw/blob/master/SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc 

In old DD, **there is a special treatment for reflected volumes, so that the volumes with _refl are added to the sensitive detector catalogue.**
In DD4hep, the G4 reflected volumes were properly created, **but their names with _refl were not added to the active detectors catalogue list.**
**This resulted, further downstream in https://github.com/cms-sw/cmssw/blob/master/SimG4Core/SensitiveDetector/src/SensitiveDetector.cc#L53, to have the reflected volumes not set as active detectors.**
This issue did not appear with the TGeo trick, because the reflected volumes with TGeo, did not have the _refl names.
The reflected active volumes being not considered as sensitive detectors by G4, there were obviously no SimHit on them.

The proper list of active volumes is now fixed, and includes the active reflected volumes.

NB: The affected active volumes, which are now present on a DD4hep workflow, and which were NOT considered as active volumes until now, without the TGeo trick, are:

**In ECAL Endcap:**
reflectedG4LogicalVolumeName = EFRY_refl
-> This solves the ECAL Endcap quadrants issue.

**In ECAL Barrel:**
reflectedG4LogicalVolumeName = ebalgo:EBRY_01_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_01_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_01_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_02_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_02_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_02_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_03_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_03_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_03_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_07_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_04_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_04_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_04_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_05_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_05_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_05_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_06_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_06_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_06_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_07_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_07_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_08_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_08_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_08_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_09_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_09_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_09_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_10_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_10_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_10_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_11_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_11_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_11_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_12_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_12_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_12_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_13_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_13_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_13_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_14_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_14_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_14_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_15_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_15_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_15_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_16_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_16_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_16_refl
reflectedG4LogicalVolumeName = ebalgo:EBRY_17_refl
reflectedG4LogicalVolumeName = ebalgo:EATJ_17_refl
reflectedG4LogicalVolumeName = ebalgo:EAPD_17_refl

**In CSC:**
reflectedG4LogicalVolumeName = csc:ME1A_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME11_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME12_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME13_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME21_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME22_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME31_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME32_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME41_ActiveGasVol_refl
reflectedG4LogicalVolumeName = csc:ME42_ActiveGasVol_refl

NB 2: PR quickly done here, might push a few commits to clean it.

@ianna @civanch @bsunanda @cvuosalo

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by ghugo83** [comment]
Ecal Endcap quadrants are back, and this time with no TGeo trick! This is using the DD4hep reflected volumes.

ZMM_10events_withNoTGeoTrick

File: SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc [hunk 1]
```diff
@@ -56,6 +57,19 @@ G4VPhysicalVolume *DDG4Builder::BuildGeometry(SensitiveDetectorCatalog &catalog)
     auto fff = it.first->GetName();
     catalog.insert({sClassName.data(), sClassName.size()}, {sROUName.data(), sROUName.size()}, fff);
 
+    auto reflectedG4LogicalVolumeName = fff + "_refl";
+    bool hasReflectedVolumeInStore = false;
+    G4LogicalVolumeStore *theStore = G4LogicalVolumeStore::GetInstance();
+    for (const auto &lv : *theStore) {
+      if (lv->GetName().find(reflectedG4LogicalVolumeName) != std::string::npos) {
+        hasReflectedVolumeInStore = true;
```
Comment (cvuosalo) [inline SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc:65]: Shouldn't you add after this line:
```
break;
```

**Comment by cvuosalo** [comment]
@ghugo83 How many logical volumes are contained in the `G4LogicalVolumeStore`? Does searching through the store one-by-one for each reflected volume take a long time? Maybe make one pass through the store?

File: SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc (hunk 1 above)
Comment (ghugo83) [reply SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc:65]: yes need to clean style a bit, will slightly improve.

Comment (ghugo83) [reply SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc:65]: But for now I was focused on fixing the issue, there was no point improving performance on wrong solutions.

Comment (ghugo83) [reply SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc:65]: break is a good point, I will add it

**Comment by ghugo83** [comment]
> @ghugo83 How many logical volumes are contained in the `G4LogicalVolumeStore`? Does searching through the store one-by-one for each reflected volume take a long time? Maybe make one pass through the store?

The overall complexity is O(number of active non-reflected logical volumes x number of logical volumes).
Please note that these are logical volumes, not placed volumes (way less).

This PR has no effect on the overall complexity of the initialization phase, because there is already an existing O(number of active non-reflected logical volumes x number of logical volumes) complexity from a loop just after anyway.

It is here: https://github.com/cms-sw/cmssw/blob/master/SimG4Core/SensitiveDetector/src/SensitiveDetector.cc#L49 
AssignSD has complexity O(number of logical volumes), and is called O(number of logical volumes) times from
 https://github.com/cms-sw/cmssw/blob/master/SimG4Core/SensitiveDetector/src/SensitiveDetector.cc#L36

This PR has same complexity, hence no observable effect on complexity overall.

**Comment by ghugo83** [comment]
The time spent at the creation of the SensitiveCatalogue is overall negligible anyway, it is not a bottleneck AND additionally, it is only called once :) 
Way more time is spent elsewhere (huge numbers of orders of magnitude more) at initialization :)

**Comment by civanch** [comment]
Number of logical volumes is ~4000 in Run-2.

**Comment by civanch** [comment]
+1

**Comment by civanch** [comment]
urgent

**Comment by qliphy** [comment]
+1

**Comment by cvuosalo** [comment]
Did we merge the version without `break;`?

**Comment by civanch** [comment]
@ghugo83 , @cvuosalo , sorry, I have signed it too fast - "break;" will speed-up the code. Let us add in the next PR.

File: SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc (hunk 1 above)
Comment (ghugo83) [reply SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc:65]: @cvuosalo Yep I remember what happened now, actually I wanted to use a std::find, but I did not know what was the type of G4LogicalVolumeStore, so just made the easy loop, to first see if the fix was working anyway.
Just looked it up, G4LogicalVolumeStore is a std::vector<G4LogicalVolume*>.
I have just used a std::find_if in https://github.com/cms-sw/cmssw/pull/32241/ , it is equivalent to a loop and break, as of course stops at first occurrence.

**Pull request merged**